### ISSUE TITLE
Should we be allowed to multi-reference keys through did documents?

### ISSUE THREAD

**Issue by kdenhartog** [w3c/did-core#878]
Original issue was authored here: https://github.com/w3c/did-core/issues/301 Migrating the issue to here because this is where it should be properly addressed.

Imagine we've got the following did document structures:

```
{
  id: "did:example:conglomerate",
  publicKey: [
  {
    id: "did:example:conglomerate#key1",
    type: "EddsaVerificationKey2018",
    base58: "a-base58-encoded-key",
  }],
  assertionMethod: ["did:example:business-unit#key2", "did:example:conglomerate#key1"]
}
```

and did:example:business-unit did Document looks like this:
```
{
  id: "did:example:business-unit",
  publicKey: [
  {
    id: "did:example:business-unit#key1",
    type: "EddsaVerificationKey2018",
    base58: "a-base58-encoded-key",
  },
  {
    id: "did:example:business-unit#key2",
    type: "DidUrlKeyReference"
    reference: "did:example:sub-division#key1"
  }
],
  assertionMethod: [ "did:example:business-unit#key2", "did:example:business-unit#key1"]
}
```

and did:example:sub-division did doc looks like this:

```
{
  id: "did:example:sub-division",
  publicKey: [
  {
    id: "did:example:sub-division#key1",
    type: "EddsaVerificationKey2018",
    base58: "a-base58-encoded-key",
  }
],
  assertionMethod: [ "did:example:sub-division#key1"]
}
```

This is essentially the same as this pseudocode:
```
global var1 = memoryLocation(0x11111111)

callerFunction() => {
const booleanResult = function1(var1); // should return false because function1 negates it
expect(booleanResult).toBe(false);
}

function1(fnVar1) => {
  !function2(fnVar1);
};

function2(fnVar2) => {
 return fnVar2 === var1; // should return true because the memory locations match
}
```

The use case where I'm imaging this would be used is did:example:conglomerate is a business conglomerate with `did:example:business-unit` being a business unit of `did:example:conglomerate`. And `did:example:sub-division` is a sub-business division of business unit, `did:example:business-unit`.

What this means is that `did:example:bussiness-unit` can sign on behalf of `did:example:conglomerate` without `did:example:conglomerate` not needing to directly reference every key managed by it's sub-divisions of it's various business units.

However, there's an implicit trust relationship between `did:example:conglomerate` and it's reference to `did:example:business-unit#key2`. For example, an attacker could compromise `did:example:business-unit` and point `#key2` to `did:example:malicious-attacker#key1` and now malicious-attacker key1 can be used to sign on behalf of `did:example:conglomerate`. 

This is similar to booleanResults expected function1 contract that it will return false because it's negating the result of function2. If it were to stop negating function2, then the logical expectation in the callerFunction would fail.

Should this be allowed? 

I tend to lean towards the answer is "No, it shouldn't be allowed" because the implicit trust relationship will be abused at some point and keys will be able to falsely represent did subjects. If others agree with me, then we don't need to make any changes to the specification and this can be close. However, the flip side is multi referencing keys enables easier management of did-documents which can be a very powerful feature. I'm curious, where do others land on the topic?

@dlongley or @msporny I suspect you guys may have thought through this case already because I see JSON-LD examples that do the direct referencing use cases in the spec.

**Comment by ankurdotb** [comment]
> However, there's an implicit trust relationship between did:example:conglomerate and it's reference to did:example:business-unit#key2. For example, an attacker could compromise did:example:business-unit and point #key2 to did:example:malicious-attacker#key1 and now malicious-attacker key1 can be used to sign on behalf of did:example:conglomerate.

In case if `did:example:business-unit#key2` is compromised:
1. The compromise is detected before a malicious attacker can use it, and the conglomerate/business unit updates the DID document to rotate keys to a different one.
2. The compromise is not detected before a malicious attacker uses it. In this scenario, the conglomerate/business unit will likely still rotate keys, and then maybe suspend/revoke credentials or connections established from the time the key was compromised.

Realistically I think the other bit that plays a part here is what's listed as the controller keys, and if there's more than 1. If the DID method allowed document updates with signoff from 1-of-many, the first thing I'd do as an attacker is to rotate keys away to something only I control and lock the conglomerate/business unit out. If the DID method treats multiple controllers as "all need to sign" or m-of-n need to sign, then while an attacker might be able to take an action (like issuing credentials) due to the compromised key being listed in `assertionMethod`, it won't be able to at least update the parent DID Document to a state where the conglomerate can't recover it.

(I hope I'm making sense.)

**Comment by msporny** [comment]
> I hope I'm making sense.

Yes, certainly -- what you're saying is aligned with where I think most of us are in our thinking these days.

M-of-N is still not broadly deployed, and probably does need to be for high-assuance, DLT-based DIDs. did:web doesn't necessarily need it as long as you trust the domain/website controller to not be compromised. There are nuances here, and the security risks don't seem to be so severe that we'd remove the capability. I expect we'll learn more about remote keys in DID Documents over the next several years.
